### PR TITLE
add-route example

### PR DESCRIPTION
**Pull request by shimonewman**
## What does this PR do?

Add route description

### CODE REVIEW

File: docs/tutorials/add-route.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
+# Adding route 🛡
+
+This document is part of the Appwrite contributors' guide. Before you continue reading this document make sure you have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/master/CODE_OF_CONDUCT.md) and the [Contributing Guide](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md).
+
+### 1. Alias
+Setting an alias  is used to permit access the route from an alias url as well,
+second parameter is used to push default values to the route.
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:7]:
```suggestion
The first parameter specifies the alias URL, the second parameter specifies default values for route parameters. 
```

File: docs/tutorials/add-route.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+```php
+App::post('/v1/storage/buckets/:bucketId/files')
+    ->alias('/v1/storage/files', ['bucketId' => 'default'])
+```
+
+### 2. desc
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:13]:
```suggestion
### 2. Description
```

File: docs/tutorials/add-route.md [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+Used as an  abstract description of the route.
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:14]:
```suggestion
Used as an abstract description of the route. 
```

File: docs/tutorials/add-route.md [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+```php
+App::init()
+->groups(['api'])
+->action(
+  some code.....
+);
+```
+
+
+### 4. The labels mechanism
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:36]:
```suggestion
### 4. The Labels Mechanism
```

File: docs/tutorials/add-route.md [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+Labels are very strait forward and easy to use and understand, but in the same time very robust.
+Labels are passed from the controllers route, and used to pick up key value pairs to be handled in a centralized place
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:38]:
```suggestion
Labels are passed from the controllers route and used to pick up key-value pairs to be handled in a centralized place
```

File: docs/tutorials/add-route.md [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+along the road.
+Labels can be used to pass a pattern in order to be replaced in the other end.
+Appwrite uses different labels to achieve different things, for example:
+
+### Scope
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:43]:
```suggestion
#### Scope
```

These subheadings under labels mechanism should be one level lower

File: docs/tutorials/add-route.md [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+* scope - Defines the route permissions scope.
+
+```php
+App::post('/v1/storage/buckets/:bucketId/files')
+    ->label('scope', 'files.write')
+```
+
+### Audit
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:51]:
```suggestion
#### Audit
```

File: docs/tutorials/add-route.md [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+
+```php
+App::post('/v1/account/create')
+->label('audits.event', 'account.create')
+->label('audits.resource', 'user/{response.$id}')
+->label('audits.userId', '{response.$id}')
+```
+
+### Sdk
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:64]:
```suggestion
#### SDK
```

File: docs/tutorials/add-route.md [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+* sdk.auth - Array of authentication types is passed in order to impose different authentication methods in different situations.
+* sdk.namespace - Refers to the route namespace.
+* sdk.method - Refers to the sdk method that needs to called.
+* sdk.description - Description of the route, using md file format.
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:68]:
```suggestion
* sdk.description - Description of the route,using markdown format.
```

File: docs/tutorials/add-route.md [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+->label('sdk.namespace', 'account')
+->label('sdk.method', 'createJWT')
+->label('sdk.description', '/docs/references/account/create-jwt.md')
+->label('sdk.response.code', Response::STATUS_CODE_CREATED)
+->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+->label('sdk.response.model', Response::MODEL_JWT)
+```
+
+### Cache
+* cache - When set to ture, signal the use of file cache.
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:84]:
```suggestion
* cache - When set to true, signal the use of file cache. 
```

File: docs/tutorials/add-route.md [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+* cache.resource - Identifies the cached resource.
+
+```php
+App::get('/v1/storage/buckets/:bucketId/files/:fileId/preview')
+->label('cache', true)
+->label('cache.resource', 'file/{request.fileId}')
+```
+
+### Abuse
+* abuse-key - Specifies routes uniq abuse key.
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:94]:
```suggestion
* abuse-key - Specifies routes unique abuse key.
```

File: docs/tutorials/add-route.md [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+```php
+App::post('/v1/storage/buckets/:bucketId/files')
+    ->label('abuse-key', 'ip:{ip},method:{method},url:{url},userId:{userId}')
+    ->label('abuse-limit', APP_LIMIT_WRITE_RATE_DEFAULT)
+    ->label('abuse-time', APP_LIMIT_WRITE_RATE_PERIOD_DEFAULT)
+```
+
+### Events
+* event - A pattern that is associated with the route in behalf of realtime messaging.
+  Placeholders marked as [] are parsed and replaced with their real values.
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:107]:
```suggestion
  Placeholders marked as `[]` are parsed and replaced with their real values.
```

File: docs/tutorials/add-route.md [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+
+```php
+App::post('/v1/storage/buckets/:bucketId/files')
+    ->label('event', 'buckets.[bucketId].files.[fileId].create')
+```
+
+### Usage
+* usage.metric - .
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:115]: I think the descriptions here are missing 😜

File: docs/tutorials/add-route.md [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+* usage.params - .
+```php
+App::post('/v1/storage/buckets/:bucketId/files')
+    ->label('usage.metric', 'files.{scope}.requests.create')
+    ->label('usage.params', ['bucketId:{request.bucketId}'])
+```
+
+### 5. Param
+As the name applies param() is used to setting up a request parameter.
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:124]:
```suggestion
As the name implies, `param()` is used to define a request parameter.
```

File: docs/tutorials/add-route.md [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+* Is the route optional
+* An array of injections
+```php
+App::get('/v1/account/logs')
+   ->param('queries', [], new Queries(new Limit(), new Offset()), 'Array of query strings generated using the Query class provided by the SDK. [Learn more about queries](https://appwrite.io/docs/databases#querying-documents). Only supported methods are limit and offset', true)
+```
+
+
+### 6. inject
+inject is used to inject dependencies pre bounded to the app.
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:140]:
```suggestion
inject is used to inject dependencies pre-bounded to the app.
```

File: docs/tutorials/add-route.md [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+
+```php
+App::post('/v1/storage/buckets/:bucketId/files')
+->inject('user')
+```
+
+In the example above user object is injected to the route pre bounded using App::setResource().
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md:147]:
```suggestion
In the example above, the user object is injected into the route pre-bounded using `App::setResource()`.
```

File: docs/tutorials/add-route.md [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+App::post('/v1/account/sessions/anonymous')
+    ->action(function (Request $request) {
+    some code...
+});
+```
+
+
+
+
+/v1/databases/:databaseId/collections/:collectionId/attributes/string
```
Comment (gewenyu99) [inline docs/tutorials/add-route.md]: Are these meant to be examples?

**Review by gewenyu99 — CHANGES_REQUESTED** [review]

File: docs/tutorials/add-route.md (hunk 1 above)
Comment (gewenyu99) [reply docs/tutorials/add-route.md:7]: I think the `:bucketId` might actually be called a slug:

https://en.wikipedia.org/wiki/Clean_URL#Slug

Comment (gewenyu99) [reply docs/tutorials/add-route.md:7]: Or this name =_- I don't actually know which is best. Pick your favorite

https://swagger.io/docs/specification/describing-parameters/#path-parameters

File: docs/tutorials/add-route.md (hunk 17 above)
Comment (shimonewman) [reply docs/tutorials/add-route.md]: Probably a paste leftover (:

File: docs/tutorials/add-route.md (hunk 13 above)
Comment (shimonewman) [reply docs/tutorials/add-route.md:115]: You are right (:

File: docs/tutorials/add-route.md [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+* cache.resource - Identifies the cached resource.
+
+```php
+App::get('/v1/storage/buckets/:bucketId/files/:fileId/preview')
+->label('cache', true)
+->label('cache.resource', 'file/{request.fileId}')
+```
+
+#### Abuse
+* abuse-key - Specifies routes unique abuse key.
```
Comment (eldadfux) [inline docs/tutorials/add-route.md:94]: Would be great to list a few of the supported vars, or briefly explain how they work.

File: docs/tutorials/add-route.md [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+* abuse-limit - Specifies the number of times the route can be requested in a time frame, per route.
+* abuse-time - Specifies the time frame relevancy of the all other abuse definitions, per route.
```
Comment (eldadfux) [inline docs/tutorials/add-route.md:96]: Seconds, minutes or hours here?

File: docs/tutorials/add-route.md [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+    ->label('usage.params', ['bucketId:{request.bucketId}'])
+```
+
+### 5. Param
+As the name implies, `param()` is used to define a request parameter.
+
+param() aspects 7 parameters :
+* A key (name) 
+* A default value
+* An instance of a relevant validator class
```
Comment (eldadfux) [inline docs/tutorials/add-route.md:129]: This one can also accept a callback that returns a validator instance. Dependency injection is supported for the callback.

File: docs/tutorials/add-route.md [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+In the example above, the user object is injected into the route pre-bounded using `App::setResource()`.
+
+```php
+App::setResource('user', function() {
+some code...
+});
+```
+
+### 6. Action
+Action populates the actual routes code and has to be very clear and understandable.
```
Comment (eldadfux) [inline docs/tutorials/add-route.md:156]:
```suggestion
Action populates the actual routes code and has to be very clear and understandable. A good route stay simple and doesn't contain complex logic. An action is where we describe our business need in code, and combine different libraries to work together and tell our story.
```

**Review by eldadfux — CHANGES_REQUESTED** [review]

**Review by gewenyu99 — APPROVED** [review]
Looks good to me!

**Review by eldadfux — APPROVED** [review]

**Pull request merged**